### PR TITLE
feat(commitlint): add reusable action for commit message linting

### PR DESCRIPTION
**Pull request by sydorovdmytro**
Adds a `commitlint` composite action. First consumer is vcluster-pro (DEVOPS-1219).

The caller owns the rules: the commitlint config always comes from the calling repository, the action only decides what gets linted and when to skip. Two independent checks, `pr-title` and a `from`/`to` range, both optional and both run to completion so one push surfaces every problem.

When the calling repo pins `@commitlint/cli` in its `package.json`, that version is installed and used, so CI runs what contributors run locally.

`skip-branches` exists for branches whose commits cannot be rewritten, notably the rebase-merged `sync-from-oss` branches that replay community authorship verbatim. It is matched against `head-repository` as well as `branch`, because a fork PR branch name is chosen by its author and would otherwise be a way to opt out of linting entirely.

`test-commitlint.yaml` also runs a smoke job against the composite itself. The bats suite calls `src/action.sh` with hand-exported `INPUT_*`, so it cannot reach the `inputs:` to `env:` mapping or the `outputs:` block, where a typo is silently green.

Closes DEVOPS-1012

## Test plan

- `make test-commitlint` - 62 bats tests pass
- `shellcheck` clean on `src/action.sh` and the test mock
- `actionlint` clean, `zizmor` reports no findings
- `make check-docs` passes
- Mutation-checked each guard against a green baseline, with the harness asserting every mutation actually applied: 12 mutations, each reddens a specific test and none is a no-op
- Ran the action against vcluster-pro's real commitlint config: good title passes, a title carrying a Linear ID fails, a real commit range passes, a `sync-from-oss/*` branch skips, and the same branch name from a fork is linted rather than skipped

### CODE REVIEW

File: .github/actions/commitlint/test/action.bats [hunk 1]
```diff
@@ -0,0 +1,796 @@
...
+@test "a branch name cannot break out of the job summary code span" {
+  export INPUT_BRANCH='x`</code>**pwned**'
+  export INPUT_SKIP_BRANCHES="x*"
+  export INPUT_HEAD_REPOSITORY="loft-sh/repo"
+  export GITHUB_REPOSITORY="loft-sh/repo"
+  export INPUT_PR_TITLE="fix: something"
+
+  run -0 "$SCRIPT"
+
+  ! grep -q '</code>' "$GITHUB_STEP_SUMMARY"
```
[the remark this answers is not in the document]
Comment (sydorovdmytro) [reply .github/actions/commitlint/test/action.bats]: Fixed, and confirmed inert exactly as described: 454 is non-final, 455 is enforced.

Took a different form than the suggestion. Two of the six bare negations in this file are pipelines (`! calls | grep -q ...`), and `run ! calls | grep -q ...` pipes `run`'s output into grep rather than negating the pipeline. So this went to two helpers instead, `assert_summary_lacks` and `assert_no_call_matching`, and all six sites are converted, including the three that are currently final. Those are only enforced until someone adds a line below them, which is how these two got here.

Mutation check: weakening `scrub()` to `tr -d '\''`'\''` reddens test 30 and nothing else. The npm half of the 633 note reddens test 44 the same way.

File: .github/workflows/test-commitlint.yaml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+        with:
+          persist-credentials: false
+
+      # The bats suite calls src/action.sh directly with hand-exported INPUT_*,
+      # so it cannot reach the `inputs:` -> `env:` mapping or the `outputs:`
+      # block in action.yml. A typo in either is invisible to every one of those
+      # tests: an input silently becomes empty, so a check quietly does not run,
+      # or an output silently becomes '' for every caller's gate.
+      - name: A conforming title passes
```
[the remark this answers is not in the document]
Comment (sydorovdmytro) [reply .github/workflows/test-commitlint.yaml:48]: Added, with one correction to the snippet. This job's checkout has no `fetch-depth`, so it gets a single commit and `git rev-parse HEAD~1` fails there. Added `fetch-depth: 2` to the checkout alongside the scenario.

Everything else is as suggested, including `pass || fail` rather than a specific verdict, and the note about `working-directory` pointing at the fixture while the revisions resolve against the repo root holds up: `range_is_reachable` runs `git cat-file` after the `cd` and the fixture directory is inside this repository.

Green on a470cf1.

On the rest of the uncovered inputs, `config-path` and `commitlint-version` are not closed. Both fail loudly if mis-mapped rather than going quietly green, which is the property that made `from` worth a scenario.

**Comment by sydorovdmytro** [comment]
Both blocking findings were real and are fixed in a470cf1, replies on the threads. On the quality notes:

**Applied**

- `action.yml:86` renovate annotation. Verified the customManager regex in `renovate.json` matches it and captures `19.8.1` with no quotes, same shape as `govulncheck`. The default had to lose its quotes for that, as noted.
- `test/action.bats:633`. Covered by the same conversion as the blocking one, all six bare negations in the file now go through helpers.
- `src/action.sh:180` capitalisation.

**Declined: `not-requested` as a distinct output value (`action.yml:96`)**

This would reopen something an earlier review round closed. `pr-title` defaults to empty, so "the caller never requested this check" and "the caller requested it and the expression resolved to empty" are indistinguishable to the action. A `not-requested` value gets emitted for both, and the gate it enables, `!= 'pass' && != 'not-requested'`, is then green on a mis-wired input that linted nothing.

The gate the README documents, `skipped != 'true' && result != 'pass'`, fails correctly there, because the `skipped` output is `true` only when `skip-branches` matched. The overload in the result enum is real, but the disambiguator is already a separate output rather than a missing one.

**Declined: declining the skip when `head-repository` is omitted (`src/action.sh:131`)**

Two reasons.

The trust model already concedes the fork case, and this bypass is not stronger than the one it concedes. A fork controls the checkout, so it can weaken `commitlint.config.js` or commit a `node_modules/.bin/commitlint` that exits 0 and get a pass without touching the branch name. Naming a branch after an exempt pattern reaches the same place by a shorter route. Closing the short route while the long one stays open buys close to nothing.

And failing toward linting would break callers that have no fork exposure to begin with. On `push` there is no `pull_request.head.repo` to pass, and a push is always to a branch in the repository, so a `push`-triggered caller with `skip-branches` would find its exempt branches linted with no way to say otherwise short of hardcoding `github.repository`.

Reading the head repo from `GITHUB_EVENT_PATH` avoids both objections and is the better version of this if it gets revisited. It was not worth the jq dependency and event-shape knowledge for the size of the gap, given the first point.

File: .github/actions/commitlint/test/action.bats [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+@test "commitlint's own echo of the title cannot forge a workflow command" {
+  # commitlint prints the message back, so its stdout is as hostile as the input.
+  export INPUT_PR_TITLE="fix: something"
+  export COMMITLINT_MOCK_STDOUT=$'validating\r::stop-commands::abc'"$PERCENT_PAYLOAD"
+
+  run -0 "$SCRIPT"
+
+  assert_no_forged_commands
+  assert_not_on_command_line "$PERCENT_PAYLOAD"
+}
```
[the remark this answers is not in the document]
Comment (sydorovdmytro) [reply .github/actions/commitlint/test/action.bats:404]: Confirmed by mutation, all four channels. No-oping the title print, the skip-path branch print, the npm print or the fork-name print each left the whole suite green.

Fixed in 939d095. Every forgery test now carries a positive assertion alongside the negative ones, and the fourth channel has a test of its own:

| mutation | tests red before | after |
|---|---|---|
| delete `print_untrusted "$PR_TITLE"` | 0 | 4 |
| delete `print_untrusted "branch: ..."` (skip path) | 0 | 1 |
| delete the npm output print | 0 | 2 |
| delete `print_untrusted "fork: ..."` | 0 | 2 |
| delete the skip job-summary line | 0 | 2 |

The job-summary test had the same shape and was not in the list: `assert_summary_lacks` holds on an empty file, so deleting the summary line entirely was green too. It now asserts the line exists as well.

Comment on the first one says why, so the next person adding a forgery test copies the pair rather than just the negative half.

File: .github/workflows/test-commitlint.yaml [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        id: pass
+        uses: ./.github/actions/commitlint
+        with:
```
[the remark this answers is not in the document]
Comment (sydorovdmytro) [reply .github/workflows/test-commitlint.yaml:51]: You are right and my earlier reply on the other thread was wrong on both counts. Neither fails loudly.

`config-path`: the flag is dropped and discovery finds `commitlint.config.js` in the same fixture directory every scenario already runs in, so the mapping can break and nothing moves.

`commitlint-version`: `${INPUT_COMMITLINT_VERSION:-19.8.1}` restores the manifest default, so a caller's pin is ignored with no signal at all.

Both closed in 939d095.

For `config-path`, took the suggestion: added `test/fixture/strict.config.js` with `scope-empty: [2, 'always']` and a scenario pointing at it with a scoped title, asserting `fail`. Verified against a real `@commitlint/cli@19.8.1` that the same title exits 0 under `commitlint.config.js` and 1 under `strict.config.js`, so the assertion follows the input rather than the fallback.

For `commitlint-version`, the only effect visible from outside the step is whether the fetch succeeds, so the scenario pins `0.0.0-does-not-exist` and asserts `pr-title-result=error`. A broken mapping falls back to 19.8.1, which works, and the assertion fails. It also covers the composite's never-started path, which nothing else here did.

Both green on 939d095.

File: .github/actions/commitlint/src/action.sh [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+# would send them chasing a problem they do not have, so a commitlint that
+# never ran is reported differently from one that ran and found problems.
+#
+# Which exit codes mean "the message has problems" depends on --strict: without
+# it commitlint returns 1, with it 2 for warnings and 3 for errors and never 1.
+#
+# That makes exit 1 unambiguous under --strict: measured against 19.8.1, an
+# unresolvable `extends` and a --from revision missing from a shallow clone both
+# exit 1 there, so 1 is exclusively a setup failure and is deliberately not in
+# the strict list.
```
[the remark this answers is not in the document]
Comment (sydorovdmytro) [reply .github/actions/commitlint/src/action.sh]: Reproduced against a real `@commitlint/cli@19.8.1` before changing anything:

| stdin | flags | exit |
|---|---|---|
| `' '` | `--strict` | 1 |
| `' '` | none | 1 |
| `'bugfix: nope'` | `--strict` | 3 |
| `'bugfix: nope'` | none | 1 |
| `'fix: ok'` | either | 0 |

So the comment was wrong and the classifier inverted the diagnostic for a reachable input. Fixed in 939d095 with the guard you suggested: a title that trims to nothing is reported as `fail` naming the blank message, and commitlint is never invoked, so its usage-exit cannot reach the classifier in either mode. New test asserts the result and that no commitlint call was made; reverting the guard reddens it.

Also corrected the comment itself. It now says exit 1 under `--strict` means commitlint never judged a message, names the `extends` case as the remaining one, and records that the shallow-clone case never reaches the classifier because `range_is_reachable` catches it first, which was your second point.

File: .github/actions/commitlint/src/action.sh [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+    log "Branch matches skip-branches; nothing linted."
+    print_untrusted "branch: ${BRANCH}"
+    print_untrusted "skip-branches: ${SKIP_BRANCHES}"
+    emit_output skipped true
+    emit_output pr-title-result skipped
+    emit_output commits-result skipped
+    emit_summary "Commitlint skipped: branch \`$(scrub "$BRANCH")\` is exempt."
+    return 0
+  fi
+  emit_output skipped false
```
[the remark this answers is not in the document]
Comment (sydorovdmytro) [reply .github/actions/commitlint/src/action.sh]: Confirmed and fixed in 2d7c2c8. `$GITHUB_OUTPUT` is inherited, append-only, last-write-wins, and `skipped` was the one output the script wrote before anything spawned.

All three outputs now go through a single `emit_results`, called only from exit paths that run after any child could have started, and nothing writes them earlier. The skip path still writes `true`, and nothing has spawned by then.

Test added with a planted `node_modules/.bin/commitlint` that rejects the title and appends `skipped=true`. It asserts the forged line is really in the file, then that the resolved value is still `false`. Restoring the exact pre-fix shape (drop `skipped` from `emit_results`, write it early instead) reddens that test and nothing else.

The `$GITHUB_STEP_SUMMARY` half is real too and is not fixed: a child can append whatever it likes there and no sanitizer in this script can stop it. Not going after it here, since unlike the output it does not feed a gate, but it belongs in the README's trust model rather than being left implicit.

File: .github/actions/commitlint/test/action.bats [hunk 7]
```diff
@@ -0,0 +1,859 @@
...
+@test "reports a whitespace-only title as a message problem, not a CI problem" {
+  export INPUT_PR_TITLE="   "
+  export INPUT_FAIL_ON_WARNINGS="true"
+
+  run -1 "$SCRIPT"
+
+  [ "$(output_value pr-title-result)" = "fail" ]
+  [[ "$output" == *"whitespace only"* ]]
+  # commitlint must never see it, or its usage-exit lands back in the classifier.
+  assert_no_call_matching "commitlint"
```
[the remark this answers is not in the document]
Comment (sydorovdmytro) [reply .github/actions/commitlint/test/action.bats]: Right on every point. I shipped a red test, and the mutation check I reported alongside it could not have been distinguishing anything.

Two separate mistakes on my side. The assertion was wrong for the reason you give: `verify_commitlint_runs` probes with `--version` and the probe is logged, so a bare `commitlint` substring always matches. It now asserts the real invariant, that nothing was handed a message to judge, via a helper that filters `--version` lines out before looking for a lint call. Reverting the whitespace guard reddens it.

The worse mistake is how I verified last round. My harness counted total failures rather than new ones against a green baseline, so the pre-existing red was folded into every count I quoted. Two of the seds also targeted a 2-space indent where the file has 4, matched nothing, and read as "guard verified" while changing not a byte. The skip-path branch print was one of those, so that channel was never actually covered in either direction. The harness now refuses to run unless the baseline is green and aborts a mutation that does not match exactly once. Twelve mutations, each reddens a specific test, no no-ops.

**On why CI was green: the bats job has never run a test.** `bats-core/bats-action@4.0.0` has no `tests` input. It installs bats and stops. An unknown input is a warning, not an error, so the job passes having executed nothing:

```
##[warning]Unexpected input(s) 'tests', valid inputs are ['bats-install', 'bats-version', ...]
```

Fixed here by installing with that action and then running `bats` in a step of its own, matching `test-release-notification.yaml`, which is the only workflow in the repo that already did it that way. The job now reports 62 tests where it reported none.

This is not local to this PR: 23 of the 24 `test-*.yaml` workflows pass the same non-existent input, so those suites have never run in CI either. Raising that separately rather than widening this PR.

**Comment by sydorovdmytro** [comment]
Both blocking findings fixed in 2d7c2c8, details on the threads. The PR-level note and the quality note are handled too, and one of them turned out to be the tip of something repo-wide.

**Test plan count** — corrected, and the underlying reason is worse than a stale number. The suite is 62 tests and one of them was red at the previous HEAD. My verification last round was unsound in two ways: the mutation harness counted total failures instead of new ones against a green baseline, so the pre-existing red was folded into every figure I quoted, and two of the seds targeted the wrong indent, matched nothing, and reported "guard verified" while changing not a byte. Rebuilt so it refuses to run on a non-green baseline and aborts any mutation that does not match exactly once. Twelve mutations, each reddens a specific test, no no-ops.

**`19.8.1` pinned twice** — right that it was live rather than dead, and the fix went a step further than deduplication. `src/action.sh` no longer carries a fallback at all: `action.yml` holds the pin Renovate manages, and an empty value is handled where it is used, in `resolve_commitlint`, which warns and resolves the latest rather than silently running a version nobody chose. That also removes the drift you flagged instead of moving it. The third copy in the root `README.md` now points at `action.yml` rather than restating the number.

**Why CI was green on a red suite** — this is the part worth reading. The bats job has never run a test. `bats-core/bats-action@4.0.0` has no `tests` input; it installs bats and stops, and an unknown input is a warning rather than an error:

```
##[warning]Unexpected input(s) 'tests', valid inputs are ['bats-install', 'bats-version', ...]
```

Fixed here by invoking `bats` in a step of its own, matching `test-release-notification.yaml`, the only workflow in the repo that already did it that way. The job now reports 62 tests where it reported none.

It is not local to this PR. 22 of the 24 bats workflows pass the same non-existent input, so **752 tests have never executed in CI** and every green check on those actions has been vacuous since the workflows were written. Filed as DEVOPS-1344 rather than widened into this PR. All 22 suites do pass locally, so the fix is expected to be mechanical, but it needs its own change and its own first run.
